numpy/core/tests/test_shape_base.py

```diff
@@ -391,39 +392,32 @@ def test_stack():
     assert_array_equal(result, np.array([0, 1, 2]))
 
 
-# See for more information on how to parametrize a whole class
-# https://docs.pytest.org/en/latest/example/parametrize.html#parametrizing-test-methods-through-per-class-configuration
-def pytest_generate_tests(metafunc):
-    # called once per each test function
-    if hasattr(metafunc.cls, 'params'):
-        arglist = metafunc.cls.params
-        argnames = sorted(arglist[0])
-        metafunc.parametrize(argnames,
-                             [[funcargs[name] for name in argnames]
-                              for funcargs in arglist])
-
-
-# blocking small arrays and large arrays go through different paths.
-# the algorithm is triggered depending on the number of element
-# copies required.
-# We define a test fixture that forces most tests to go through
-# both code paths.
-# Ultimately, this should be removed if a single algorithm is found
-# to be faster for both small and large arrays.s
-def _block_force_concatenate(arrays):
-    arrays, list_ndim, result_ndim, _ = _block_setup(arrays)
-    return _block_concatenate(arrays, list_ndim, result_ndim)
-
-
-def _block_force_slicing(arrays):
-    arrays, list_ndim, result_ndim, _ = _block_setup(arrays)
-    return _block_slicing(arrays, list_ndim, result_ndim)
-
-
 class TestBlock(object):
-    params = [dict(block=block),
-              dict(block=_block_force_concatenate),
-              dict(block=_block_force_slicing)]
+    @pytest.fixture(params=['block', 'force_concatenate', 'force_slicing'])
+    def block(self, request):
+        # blocking small arrays and large arrays go through different paths.
+        # the algorithm is triggered depending on the number of element
+        # copies required.
+        # We define a test fixture that forces most tests to go through
+        # both code paths.
+        # Ultimately, this should be removed if a single algorithm is found
+        # to be faster for both small and large arrays.
+        def _block_force_concatenate(arrays):
+            arrays, list_ndim, result_ndim, _ = _block_setup(arrays)
+            return _block_concatenate(arrays, list_ndim, result_ndim)
+
+        def _block_force_slicing(arrays):
+            arrays, list_ndim, result_ndim, _ = _block_setup(arrays)
+            return _block_slicing(arrays, list_ndim, result_ndim)
+
+        if request.param == 'force_concatenate':
+            return _block_force_concatenate
+        elif request.param == 'force_slicing':
+            return _block_force_slicing
+        elif request.param == 'block':
+            return block
```

Comment:
Note you could also achieve this via the `ids` parameter to `fixture`:
```
@pytest.fixture(
    params=[block, _block_force_concatenate, _block_force_slicing],
    ids=['block', 'force_concatenate', 'force_slicing']
)
```
Which has the benefit of not needing the final `else`.
Not really a big deal either way.

Reply:
that else statement is more a test on the test. It checks that I know how to use request (which I didn't).